### PR TITLE
enhancement(kafka sink, codecs): Integrate `encoding::Encoder` with `kafka` sink

### PR DESCRIPTION
**Pull request by pablosichert**
Part of #9459.

### CODE REVIEW

File: src/sinks/kafka/config.rs [hunk 1]
```diff
@@ -20,12 +26,34 @@ use crate::{
 
 pub(crate) const QUEUED_MIN_MESSAGES: u64 = 100000;
 
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Migrator;
+
+impl EncodingConfigMigrator for Migrator {
```
Comment (fuchsnj) [inline src/sinks/kafka/config.rs:32]: Would it be reasonable to make some "standard" migrators, and use those? I don't see anything that seems very specific to this sink here.

Comment (pablosichert) [reply src/sinks/kafka/config.rs:32]: Yes: https://github.com/vectordotdev/vector/pull/12136#discussion_r846324039 😁

File: src/sinks/kafka/config.rs [hunk 2]
```diff
@@ -20,12 +26,34 @@ use crate::{
...
+    type Codec = StandardEncodings;
+
+    fn migrate(codec: &Self::Codec) -> (Option<FramingConfig>, SerializerConfig) {
+        let framing = match codec {
+            StandardEncodings::Text | StandardEncodings::Json => None,
+            StandardEncodings::Ndjson => Some(NewlineDelimitedEncoderConfig::new().into()),
+        };
```
Comment (lukesteensen) [inline src/sinks/kafka/config.rs:39]: Does having a framer in a message-oriented sink like kafka make sense?

Comment (jszwedko) [reply src/sinks/kafka/config.rs:39]: Yeah, I had this same thought. It seems very unlikely to me that someone would want to use anything other than message framing for any message-based sink (Kafka, UDP, AWS SQS, GCP Pub Sub, etc.). I'd support restricting it to message-based framing until we get a user with a use-case for something else.

Comment (pablosichert) [reply src/sinks/kafka/config.rs]: Agree, that makes more sense. Have a PR up for the necessary support structures: https://github.com/vectordotdev/vector/pull/12156.

**Review by jszwedko — APPROVED** [review]

**Review by lukesteensen — APPROVED** [review]

**Pull request merged**